src/devices/ShiftRegister/ShiftRegister.cs

```diff
@@ -0,0 +1,243 @@
+        {
+            _spiDevice = spiDevice;
+            _bitLength = bitLength;
+        }
+
+        /// <summary>
+        /// GPIO controller.
+        /// </summary>
+        protected GpioController GpioController => _controller;
+
+        /// <summary>
+        /// SPI device.
+        /// </summary>
+        protected SpiDevice SpiDevice => _spiDevice;
+
+        /// <summary>
+        /// Bit length across all connected registers.
+        /// </summary>
+        public int BitLength => _bitLength;
+
+        /// <summary>
+        /// Reports if shift register is connected with SPI.
+        /// </summary>
+        public bool UsesSpi => _spiDevice is object;
+
+        /// <summary>
+        /// Reports if shift register is connected with GPIO.
+        /// </summary>
+        public bool UsesGpio => _controller is object;
+
+        /// <summary>
+        /// Shifts zeros.
+        /// Will dim all connected LEDs, for example.
+        /// Assumes register bit length evenly divisible by 8.
+        /// Supports GPIO controller or SPI device.
+        /// </summary>
+        public void ShiftClear()
+        {
+            if (_bitLength % 8 > 0)
+            {
+                throw new ArgumentNullException($"{nameof(ShiftClear)}: Only supported for registers with bit lengths evenly divisible by 8.");
+            }
+
+            for (int i = 0; i < _bitLength / 8; i++)
+            {
+                ShiftByte(0b_0000_0000);
+            }
+        }
+
+        /// <summary>
+        /// Writes PinValue value to storage register.
+        /// This will shift existing values to the next storage slot.
+        /// Does not latch.
+        /// Requires use of GPIO controller.
+        /// </summary>
+        public void ShiftBit(PinValue value)
+        {
+            if (_controller is null || _pinMapping.SerialDataInput < 0)
+            {
+                throw new ArgumentNullException($"{nameof(ShiftBit)}: GpioController was not provided or {nameof(_pinMapping.SerialDataInput)} not mapped to pin");
+            }
+
+            // writes value to serial data pin
+            _controller.Write(_serial, value);
+            // data is written to the storage register on the rising edge of the storage register clock
+            _controller.Write(_clock, 1);
+            // values are reset to low
+            _controller.Write(_serial, 0);
+            _controller.Write(_clock, 0);
+        }
+
+        /// <summary>
+        /// Shifts a byte -- 8 bits -- to the storage register.
+        /// Assumes register bit length evenly divisible by 8.
+        /// Pushes / overwrites any existing values.
+        /// Latches by default.
+        /// </summary>
+        public void ShiftByte(byte value, bool latch = true)
+        {
+            if (_spiDevice is object)
+            {
+                _spiDevice.WriteByte(value);
+                return;
+            }
+
+            for (int i = 0; i < 8; i++)
```

Comment:
this part looks like software SPI. Is it possible to use SoftwareSpi in a specific mode (CPHA=false, CPOL=false; or mode=0)  for this?

If so we could remove part of the abstraction like (UsesSpi, UsesGpio). BitLength can also be set on both Software and HardwareSpi (through settings)

Reply:
I'm not seeing a way to do what you are suggesting.

GPIO access is about changing one bit at a time. SPI is about updating on byte at a time. They are different patterns. This code allows you, as a secondary pattern, to update a byte at a time from GPIO.

Reply:
For example, the `Mbi5027` needs to interact with `OE`, `Sdo`, and `Le` for error detection. `OE` and `Sdo` would not be used by  `SoftwareSpi`, but `Le` would be. It would seem odd to share the same pin across these two different use cases.